appservice/message_send_checkpoint.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2021 Sumner Evans
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package appservice
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"maunium.net/go/mautrix/event"
+	"maunium.net/go/mautrix/id"
+)
+
+type MessageSendCheckpointStep string
+
+const (
+	ClientStep     MessageSendCheckpointStep = "CLIENT"
+	HomeserverStep MessageSendCheckpointStep = "HOMESERVER"
+	BridgeStep     MessageSendCheckpointStep = "BRIDGE"
+	RemoteStep     MessageSendCheckpointStep = "REMOTE"
+	CommandStep    MessageSendCheckpointStep = "COMMAND"
+)
+
+type MessageSendCheckpointStatus string
+
+const (
+	SuccesssStatus    MessageSendCheckpointStatus = "SUCCESS"
+	WillRetryStatus   MessageSendCheckpointStatus = "WILL_RETRY"
+	PermFailureStatus MessageSendCheckpointStatus = "PERM_FAILURE"
+)
+
+type MessageSendCheckpointReportedBy string
+
+const (
+	ReportedByAsmux  MessageSendCheckpointReportedBy = "ASMUX"
+	ReportedByBridge MessageSendCheckpointReportedBy = "BRIDGE"
+)
+
+type MessageSendCheckpoint struct {
+	EventID     id.EventID                      `json:"event_id"`
+	RoomID      id.RoomID                       `json:"room_id"`
+	Step        MessageSendCheckpointStep       `json:"step"`
+	Timestamp   time.Time                       `json:"timestamp"`
+	Status      MessageSendCheckpointStatus     `json:"status"`
+	EventType   event.Type                      `json:"event_type"`
+	ReportedBy  MessageSendCheckpointReportedBy `json:"reported_by"`
+	RetryNum    int                             `json:"retry_num"`
+	MessageType event.MessageType               `json:"message_type"`
+	Info        string                          `json:"info"`
+}
+
+func (cp *MessageSendCheckpoint) Send(endpoint string, asToken string) error {
+	return SendCheckpoints(endpoint, asToken, []*MessageSendCheckpoint{cp})
+}
+
+type CheckpointsJson struct {
+	Checkpoints []*MessageSendCheckpoint `json:"checkpoints"`
+}
+
+func SendCheckpoints(endpoint string, asToken string, checkpoints []*MessageSendCheckpoint) error {
+	if endpoint == "" {
+		return nil
+	}
+
+	var body bytes.Buffer
+	checkpointsJson := CheckpointsJson{Checkpoints: checkpoints}
+	if err := json.NewEncoder(&body).Encode(checkpointsJson); err != nil {
+		return fmt.Errorf("failed to encode message send checkpoint JSON: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, &body)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Authorization", "Bearer "+asToken)
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req)
```

Comment:
We should probably add the proper user agent header for these requests (from `mautrix.DefaultUserAgent`, and maybe some extra string like "checkpoint sender" like https://github.com/mautrix/go/blob/master/client.go#L107)